include/boost/decimal/detail/cmath/sqrt.hpp

```diff
@@ -24,63 +23,132 @@ namespace decimal {
 namespace detail {
 
 template <typename T>
-constexpr auto sqrt_impl(T val) noexcept
+constexpr auto sqrt_impl(T x) noexcept
     BOOST_DECIMAL_REQUIRES(detail::is_decimal_floating_point_v, T)
 {
-    constexpr T zero {0, 0};
+    const auto fpc = fpclassify(x);
 
     T result { };
 
-    if (isnan(val) || abs(val) == zero)
+    if ((fpc == FP_NAN) || (fpc == FP_ZERO))
     {
-        result = val;
+        result = x;
     }
-    else if (isinf(val))
+    else if (signbit(x))
     {
-        if (signbit(val))
-        {
-            result = std::numeric_limits<T>::quiet_NaN();
-        }
-        else
-        {
-            result = val;
-        }
+        result = std::numeric_limits<T>::quiet_NaN();
     }
-    else if (val < zero)
+    else if (fpc == FP_INFINITE)
     {
-        result = std::numeric_limits<T>::quiet_NaN();
+        result = std::numeric_limits<T>::infinity();
     }
     else
     {
-        // Loosens tolerance with the increase in digits so it's not excessively expensive
-        constexpr T epsilon = std::numeric_limits<T>::epsilon() * std::numeric_limits<T>::digits10;
-        constexpr T one { 1, 0 };
-        constexpr T half {5, -1};
-        T error = one / epsilon;
-
-        T x {};
-        if (val > one)
+        int exp10val { };
+
+        const auto gn { frexp10(x, &exp10val) };
+
+        const auto
+            zeros_removal
+            {
+                remove_trailing_zeros(gn)
+            };
+
+        const bool is_pure { static_cast<unsigned>(zeros_removal.trimmed_number) == 1U };
+
+        constexpr T one { 1 };
+
+        if(is_pure)
         {
-            // Scale down if val is large by dividing the exp by 2
-            int exp {};
-            auto sig = frexp10(val, &exp);
-            x = T{sig, exp / 2};
+            // Here, a pure power-of-10 argument gets a straightforward result.
+            // For argument 10^n where n is even, the result is exact.
+
+            const int p10 { exp10val + static_cast<int>(zeros_removal.number_of_removed_zeros) };
+
+            if (p10 == 0)
+            {
+                result = one;
+            }
+            else
+            {
+                const int p10_mod2 = (p10 % 2);
+
+                result = T { 1, p10 / 2 };
+
+                if (p10_mod2 == 1)
+                {
+                    result *= numbers::sqrt10_v<T>;
```

Comment:
Same comment about adding to docs

Reply:
I added $\sqrt{10}$, $\sqrt[3]{10}$ and $\sqrt[3]{2}$. I can document these accordingly.

Reply:
Docs updated and slightly corrected. Please Matt (@mborland) could you make sure my corrections in the ASCII-docs are OK?